### PR TITLE
Fix issues with inferred Scala classpath configurations

### PR DESCRIPTION
**Pull request by pantherdd**
### Context
This pull request fixes the following issues:
* https://github.com/gradle/gradle/issues/27226
* https://github.com/gradle/gradle/issues/27309
* https://github.com/gradle/gradle/issues/18435

As a side improvement, it also makes the `Task` interface extend `Named`, similar to the recent improvement to the `Configuration` interface.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew :scala:sanityCheck`.
- [x] Ensure that unit tests pass locally: `./gradlew :scala:test`.
- [x] Ensure that tests pass locally: `./gradlew :scala:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by pantherdd** [comment]
I built Gradle locally including my changes, and using this version resolved the issues in my earlier reproductions. 🙂
Ready for review.

Should I add any integration tests specifically for these issues, or would the existing ones and the unit tests be enough?

**Comment by pantherdd** [comment]
Could these fixes still make their way into the `8.7` release? If so, should I be targeting this PR at a different branch, e.g. [`release`](https://github.com/gradle/gradle/tree/release)?

**Comment by pantherdd** [comment]
Rebased my changes on the `release` branch and switched the PR's target to that too.

**Comment by pantherdd** [comment]
@jvandort I rebased my changes on the `release` branch and switched the PR's target to that too, based on how https://github.com/gradle/gradle/pull/28267 got merged into the [`8.7 RC2` milestone](https://github.com/gradle/gradle/milestone/270).

File: platforms/jvm/scala/build.gradle.kts
Comment (pantherdd) [inline platforms/jvm/scala/build.gradle.kts:1]: Dependencies `libs.groovy` and `libs.inject` are included transitively via `project(":core-api")`:
* https://github.com/gradle/gradle/blob/release/subprojects/core-api/build.gradle.kts#L35
* https://github.com/gradle/gradle/blob/release/subprojects/core-api/build.gradle.kts#L39

File: platforms/jvm/scala/build.gradle.kts [hunk 1]
```diff
@@ -34,9 +34,7 @@ dependencies {
     implementation(project(":process-services"))
     implementation(project(":toolchains-jvm"))
 
-    implementation(libs.groovy)
     implementation(libs.guava)
-    implementation(libs.inject)
```
Comment (pantherdd) [inline platforms/jvm/scala/build.gradle.kts:39]: Dependencies `libs.groovy` and `libs.inject` are included transitively via `project(":core-api")`:
* https://github.com/gradle/gradle/blob/release/platforms/jvm/scala/build.gradle.kts#L21
* https://github.com/gradle/gradle/blob/release/subprojects/core-api/build.gradle.kts#L35
* https://github.com/gradle/gradle/blob/release/subprojects/core-api/build.gradle.kts#L39

**Comment by ov7a** [comment]
Thank you for your contribution!

Please, move out your changes related to `Task` interface as they are not scala-specific and would require a separate consideration since it's a change in the core API.

****

> Should I add any integration tests specifically for these issues, or would the existing ones + new unit tests be enough?

Will answer to that later, after the review from the community.

> Could these fixes still make their way into the 8.7 [release](https://github.com/gradle/gradle/tree/release)? If so, should I be targeting this PR at a different branch, e.g. release?

No. The release window for 8.7 is closed. This PR can land in 8.8 if it's merged before April. 

****

@ijuma @tgodzik @dubinsky @SheliakLyr If you have time, can you take a look at this PR? As you know, we are counting on the Scala Community to help improve Gradle Scala Plugin.

**Comment by pantherdd** [comment]
@ov7a

> The release window for 8.7 is closed. This PR can land in 8.8 if it's merged before April.

Ah, too bad. I was hoping that bug fixes can still get included in the next `8.7` release candidate.
I'll rebase everything back on top of `master` then, switch the PR back to target `master`, and update the `@since` in my Javadocs to `8.8`.

> Please, move out your changes related to `Task` interface as they are not scala-specific and would require a separate consideration since it's a change in the core API.

Will do, but some of the code in this PR depends on that change. How should I represent this, so that the PR remains buildable? The idea I have:

1. Rebase commit b05254bfcdff1a011366f7dd8705a951425b8ec0 on top of `master`, and move it to a separate branch and PR targeting `master`
2. Rebase the commits remaining in this PR on top of that new branch, and keep targeting `master`

Is this the correct approach? I believe that the commit related to the changes in `Task` would still show up here initially, but disappear once that PR gets merged.

Thanks!

**Comment by ov7a** [comment]
> Will do, but some of the code in this PR depends on that change. How should I represent this, so that the PR remains buildable?

Please make this PR buildable without that change. You can create a separate PR for the `Task` changes, but please keep them independent. I understand this can result in some complications to this PR, but since the `Task` change is a core API change, it may be rejected or postponed until 9.0.

**Comment by pantherdd** [comment]
@ov7a 

> Please make this PR buildable without that change. You can create a separate PR for the `Task` changes, but please keep them independent. I understand this can result in some complications to this PR, but since the `Task` change is a core API change, it may be rejected or postponed until 9.0.

I have high hopes that it will be accepted quickly, given that it's based on a very similar recent change, and it's completely backwards-compatible. I just submitted the separate PR, if someone could take a quick glance at it (max. 2 minutes), that'd be much appreciated: https://github.com/gradle/gradle/pull/28282

With that said, I understand your concern. I can make this PR independent without a lot of effort, it's just that the error reporting won't be as descriptive without that core API improvement.

**Comment by pantherdd** [comment]
I updated the code to avoid relying on https://github.com/gradle/gradle/pull/28282, and created a separate `scala-plugin-cleanup` branch in my fork to demonstrate the envisioned end state after the other PR is merged: https://github.com/pantherdd/gradle/compare/scala-plugin...scala-plugin-cleanup

**Comment by pantherdd** [comment]
@ov7a The CI build you started failed with this error:
```
Gradle could not start your build.
> Could not create service of type BuildCacheController using .createBuildCacheController().
   > Cannot use Gradle Enterprise build cache client without specifying Gradle Enterprise server address.
```
It was the same yesterday; I thought that maybe it was just a one-off, but it seems like it isn't. My guess is that this is unrelated my changes, and it's likely some central issue with the CI setup, but just to make sure: is there anything I can/need to do to fix this?

**Comment by ov7a** [comment]
@pantherdd 
Yes, this is a known CI problem; it's unrelated to your changes, and you can ignore it.
It is fixed in the master already, so you can rebase your PR, but this action is optional.

**Comment by pantherdd** [comment]
@ov7a No worries, I'm happy to do such small chores to have a successful CI run. But unfortunately rebasing the changes on current `master` did not fix the issue.

**Comment by ov7a** [comment]
@pantherdd whoops. Sorry about that. I've notified the responsible team.

**Comment by pantherdd** [comment]
@bot-gradle test this

**Comment by pantherdd** [comment]
Most of the failures are fixed now, there are only 2 remaining:
* `GradleBuildSanityCheckConfigurationCacheSmokeTest`, which seems unrelated
* `:architecture-test:test` shows that I broke 5 rules
  * `ConfigurationCreationTest` > `configurations_are_created_with_role_based_api`
  * `ProviderMigrationArchitectureTest` > `public_api_task_properties_are_providers`
  * `ProviderMigrationArchitectureTest` > `public_api_task_file_properties_are_configurable_file_collections`
  * `PublicApiCorrectnessTest` > `public_api_methods_do_not_reference_internal_types_as_parameters`
  * `PublicApiCorrectnessTest` > `public_api_classes_do_not_extend_internal_types`

How strictly are these rules enforced, do I have to find a way to abide by all of them or is there a way to get an exception? For example:
* The `Configuration`s registered by `ScalaRuntime` would have to use the `ConfigurationRoles.RESOLVABLE_DEPENDENCY_SCOPE` as far as I understand, but there does not seem to be a way to _register_ such a `Configuration`
* `ScalaTask.getConventionMapping()` does not return a `Provider`, and I can't really change that because I'm inheriting that method from `IConventionAware`, but have to declare it myself too only to add an `@Internal` annotation on it
* `ScalaTask extends IConventionAware`, yes, but if that's a no-go, what should it do instead?
* The other method declarations of `ScalaTask` are copied from `ScalaCompile` and `ScalaDoc`, I can't change their signature either
* `ScalaJar.inspect` returns a `Stream`, which is not allowed, but what is the proper alternative that allows short-circuited filtering? Or should I throw the optimization aspect away and just return a `Collection` (or something similar)?

**Comment by dubinsky** [comment]
> If you have time, can you take a look at this PR? As you know, we are counting on the Scala Community to help improve Gradle Scala Plugin.

This pull request is, in my opinion, very helpful and well executed; I hope it gets accepted soon.

**Comment by pantherdd** [comment]
Rebased on latest `master` now that https://github.com/gradle/gradle/pull/28282 is merged

We can now remove the redundant extends from `ScalaTask` (since now `Task` itself extends `Named`).

**Comment by pantherdd** [comment]
It's only the `:architecture-test:test` rules now that are causing PR check failures, but I'd need some advice from the maintainers on what I should do about those. Detailed questions are in [my earlier comment](https://github.com/gradle/gradle/pull/28198#issuecomment-1972938677).

**Comment by ov7a** [comment]
@pantherdd In general, new code should obey the architecture rules. There might be exceptions, but you need a solid reason to get an exception for the rule. For the Scala plugin, it might be something like "it was that way before, it is already in the exception list, and it's hard to change", but it's up to the @gradle/bt-jvm team to decide.

> there does not seem to be a way to register such a Configuration

Please look at `ConfigurationContainer.resovable` and `ConfigurationContainer.dependencyScope`

> The other method declarations of ScalaTask are copied from ScalaCompile and ScalaDoc, I can't change their signature either

`getScalaClasspath()` and similar are in the exclusion list, so it's up to JVM team to decide.

> ScalaTask extends IConventionAware, yes, but if that's a no-go, what should it do instead?

Potentially you can avoid this problem by extending `SourceTask` - both `ScalaDoc` and `ScalaCompile` extend it already.

> ScalaJar.inspect returns a Stream, which is not allowed, but what is the proper alternative

You can use `Iterable` or `Collection`, yes.

**Comment by pantherdd** [comment]
@ov7a Thanks for the guidance, I was able to modify my code in a way that it now passes the `architecture-test` checks too. 👍

Notes:
* The code is still calling `project.configurations.register`, so no idea why `configurations_are_created_with_role_based_api` doesn't fail anymore. 😮
  * I can't switch to using `ConfigurationContainer.resovable` / `ConfigurationContainer.dependencyScope` because I need both roles here (if I understand them correctly).
  * I could _create_ such a configuration using `RoleBasedConfigurationContainerInternal.resolvableDependencyScopeUnlocked`, but is there a way to _register_ one instead?
* `ScalaTask` can't extend `SourceTask` because the latter is an abstract class, and `ScalaTask` needs to be an interface to keep the class inheritance tree of `ScalaCompile` and `ScalaDoc` intact. The solution I came up with was to use a runtime check whether a given `ScalaTask` instance implements `IConventionAware`, and only register a `scalaClasspath` convention if it does. This allows users to implement their own `ScalaTask` classes without the convention.
* I added `ScalaTask`'s `FileCollection` properties to the exception list (`provider-task-file-collection.txt`), pending approval of course.
* `ScalaJar.inspect` now returns an `Iterable` instead of a `Stream`. It's a bit more clunky to use, but it follows the rules. Same for the `VersionNumber` property.

**Comment by ov7a** [comment]
@pantherdd Thanks! For the remaining issues - it's up to @gradle/bt-jvm team.

**Comment by pantherdd** [comment]
> For the remaining issues - it's up to @gradle/bt-jvm team.

@ov7a The @gradle/bt-jvm tagging doesn't seem to get recognized by Github (at least it's not clickable) – does that team need to be notified about this in some way, or do they see that their assistance is required here through some task board? Thanks 🙂

**Comment by ov7a** [comment]
> @ov7a The @gradle/bt-jvm tagging doesn't seem to get recognized by Github (at least it's not clickable) – does that team need to be notified about this in some way, or do they see that their assistance is required here e.g. through a task board? Thanks 🙂

This is a private group. It's on their board, don't worry. However, this may take time for them to get to it due to other priorities. Sorry for the delay.

**Comment by jvandort** [comment]
Sorry it has taken so long for me to review this one and thank you for going through the effort of enhancing the Scala infrastructure. We certainly need improvements in this area. 

Unfortunately, this is the opposite direction we want to go in. The main issue is that this PR doubles-down on inferring scala versions from the classpath and this is a behavior we want to move away from. 

In general, we want the user to be explicit. There should be a field on the scala extension where the user declares the version of scala they want to use. This would then translate to _us_ adding the scala dependency for the user, similar to how test suites adds test framework dependencies for the user.

As it exists, we need to go through all this complicated and error-prone work to parse through the scala jar and inspect its contents. None of this should be required. The existing logic to parse file names and extract jars should all go. 

Another thing we want to avoid is exposing low-level wiring to users. Much of the public methods in `ScalaRuntime.java`, even if they were merged in, should not be exposed and accessible to the user. These are not the sorts of utilities that should be accessible from buildscripts. 

I am closing this PR since we do not want to go in this direction. Instead, we should discuss a solution more in-line with our goals on https://github.com/gradle/gradle/issues/18435, which hints at the core problem that there is no way to _declare_ the scala version. 

In general, we would want to end up with a solution that looks something like below, with no inference at all:
```
plugins {
    id("scala")
}

scala {
    runtimeVersion = "foo"
}
```

**Comment by kisssadam** [comment]
@jvandort @ljacomet I understand that you do not want to focus on this because a full rewrite is expected in later major version of Gradle, but as far as I can tell @pantherdd's PR keeps everything backward compatible.

Would it be possible to use this PR as a temporary fix for an already broken behavior in a backward compatible way before you get a chance to completely rewrite it in newer versions?

**Comment by jvandort** [comment]
This PR introduces several new APIs and deprecates several others. We want to avoid API churn like this, especially if those APIs will be replaced _again_ in the short to medium term. Any new APIs introduced should move towards explicitly declaring the scala version.

If classpath inspection can be improved without changing the API, we would consider merging that code. Otherwise effort should be focused on a solution where the version is declared.

**Comment by pantherdd** [comment]
@jvandort Thanks for the explanation of what the actual blocker is. If I understand correctly, it's the methods that are marked with `@Incubating`, as they create new public APIs that may not mature but will have to be thrown away. Based on your reply, we can get back to discussing potential solutions.

1. Before I even started working on these changes, I was told that "_[Scala Plugin is out of the team's focus for now and it's unlikely that the related issues will be addressed in the near future](https://github.com/gradle/gradle/issues/27226#issuecomment-1838829207)_". Has this changed? Are there plans now for implementing the set of (breaking) changes that you described as the long-term solution? What is the planned version of Gradle that such changes could be included in?
2. According to the [docs](https://docs.gradle.org/current/javadoc/org/gradle/api/Incubating.html), `@Incubating` means that "_the feature is currently a work-in-progress and may change at any time_". Is it a problem if such an API gets dropped after a few minor versions, before it could become stable?
3. What changes would I need to make for this _backwards-compatible_ changeset to become acceptable?
   - I could remove the `@Deprecated` annotation I added to the old APIs, meaning they would still be considered the stable solution, and my new APIs would become completely optional, used by only those who really need them. These users would know the APIs could get dropped any time because they're `@Incubating`.
   - I could even move all of the new APIs into an internal package, so they wouldn't even become `@Incubating` public APIs. Maybe we could make some exceptions, as some APIs point in exactly the direction that you're proposing (creating a Scala classpath configuration for an explicitly specified version, solving #18435).

Let me know what you think. Provided that you see some way forward, can you please reopen the PR?
Thanks.

**Comment by pantherdd** [comment]
@jvandort, @ov7a, @dubinsky What do you all think about the proposed solutions? I would really like to avoid throwing all this work out the window. And I would still like to get this into [#266](https://github.com/gradle/gradle/milestone/266), because the improvements are basically ready, I just need to know what exact modifications would make this acceptable (to avoid further wasted effort).

**Comment by big-guy** [comment]
Sorry, Justin's out right now.

There are a couple of blocking issues in the PR. We do not want to create more `Configuration`s, especially when they're used for a single purpose (e.g., like is done in `IdeaScalaConfigurer`). 

We also don't want any of the APIs added in `ScalaRuntime` to creates/configure `Configuration`s to be used for resolving Scala classpaths. It doesn't matter that they are `@Incubating`. 

I think we're trying to solve too many issues at once, so to pare it back a bit, I think we can address just https://github.com/gradle/gradle/issues/27226 

There are two ideas here:
1. Instead of looking at the file names, let's load up the given classpath in `ScalaRuntime.inferScalaClasspath` and ask what version it is. This might be slow, but we can cache or speed it up. This does't require any new APIs. 
2. Introduce a `ScalaRuntime.scalaVersion` of type `Property<String>`. When this is set, we use whatever it says. Otherwise, we fallback to the old behavior. 

(1) is doable. We don't need to support a generic way to lookup a Scala version from a jar or classpath. I think there are only 3 places that look up the Scala version:

- `ScalaRuntime` definitely needs the Scala version, but this can be contained.
- `IdeaScalaConfigurer` looks like it uses the Scala version to generate a name in IntelliJ, but I can't believe that the name is meaningful. It seems like we could just drop the version number.
- `ScalaCompileOptionsConfigurer` needs the Scala version to determine what command-line arguments to pass, but I think this can be deferred and done later. Ultimately, `ZincScalaCompilerArgumentsGenerator` needs to know the command-line argument, but I think we can figure out what the appropriate CLI argument is there if we just know the effective target.

**Pull request closed without merging**